zeebe/broker/src/main/java/io/camunda/zeebe/broker/transport/commandapi/CommandApiServiceImpl.java

```diff
@@ -76,44 +74,27 @@ protected void onActorClosing() {
   }
 
   @Override
-  public ActorFuture<Void> onBecomingFollower(final int partitionId, final long term) {
-    return removeLeaderHandlersAsync(partitionId);
-  }
-
-  @Override
-  public ActorFuture<Void> onBecomingLeader(
-      final int partitionId,
-      final long term,
-      final LogStream logStream,
-      final QueryService queryService) {
-    final CompletableActorFuture<Void> future = new CompletableActorFuture<>();
-    actor.call(
+  public ActorFuture<Void> registerHandlers(
+      final int partitionId, final LogStream logStream, final QueryService queryService) {
+    return actor.call(
         () -> {
+          // create the writer immediately so if the logStream is closed, this will throw an
+          // exception immediately
+          final var logStreamWriter = logStream.newLogStreamWriter();
           leadPartitions.add(partitionId);
           queryHandler.addPartition(partitionId, queryService);
           serverTransport.subscribe(partitionId, RequestType.QUERY, queryHandler);
-
-          final var logStreamWriter = logStream.newLogStreamWriter();
           commandHandler.addPartition(partitionId, logStreamWriter);
           serverTransport.subscribe(partitionId, RequestType.COMMAND, commandHandler);
-          future.complete(null);
         });
-    return future;
   }
 
   @Override
-  public ActorFuture<Void> onBecomingInactive(final int partitionId, final long term) {
-    return removeLeaderHandlersAsync(partitionId);
-  }
-
-  private ActorFuture<Void> removeLeaderHandlersAsync(final int partitionId) {
-    return actor.call(() -> removeLeaderHandlers(partitionId));
-  }
-
-  private void removeLeaderHandlers(final int partitionId) {
+  public ActorFuture<Void> unregisterHandlers(final int partitionId) {
     commandHandler.removePartition(partitionId);
     queryHandler.removePartition(partitionId);
     cleanLeadingPartition(partitionId);
+    return createCompletedFuture();
```

Comment:
:x: This method should be wrapped in `actor.call` or `actor.run` because it is modifying data local to this actor which are not thread-safe.